apps/emqx_resource/src/emqx_resource_buffer_worker.erl

```diff
@@ -875,6 +875,15 @@ do_call_query(_QM, _Id, _Index, _Ref, _Query, _QueryOpts, _Data) ->
         %% return `{error, {recoverable_error, Reason}}`
         EXPR
     catch
+        %% For convenience and to make the code in the callbacks cleaner an
+        %% error exception with the two following formats are translated to the
+        %% corresponding return values. The receiver of the return values
+        %% recognizes these special return formats and use them to decided if a
+        %% request should be retried.
+        error:{unrecoverable_error, Msg} ->
+            {error, {unrecoverable_error, Msg}};
+        error:{recoverable_error, Msg} ->
+            {error, {recoverable_error, Msg}};
```

Comment:
I think that resources that do want to signal a recoverable error should just return an error value (like `{error, {recoverable_error | unrecoverable_error, _}}`) instead of `error({recoverable_error, _})`, so this wouldn't be strictly necessary.  

In that case, I'd suggest we only treat that case (error values) rather than open up the possibility of raising errors with the same meaning, so we have only one way to reason and signal those intentions rather than multiple.  

What do you think?

Reply:
My thinking was that by being able to throw an exception to signal these special types of errors one could make the code in the callback cleaner. Especially if the error happen deep down in the call stack and not in the top level callback function the code will get more messy if one have to check for error return values after every function call. Also, we capture all exceptions anyway so this just add the possibility to tell the caller more precisely what type of error it is when one throws an exception. Essentially one could accomplish the same thing by wrapping the callback function in a try-catch but if this is done in several bridges there will be a lot of unnecessary repetition.  You are a more experienced Erlang programmer than me so I will let you have the final say in this.

Reply:
I see, I understand that it's currently inconvenient when we want to abort some long expression chain.  Hopefully we'll be able to use [`maybe`](https://www.erlang.org/doc/reference_manual/expressions.html#maybe) when we've fully moved to OTP 25 :smile_cat: .  

Usually, I'd say that returning error values rather than using exceptions for the flow control is easier to reason about.  But since it's a bit inconvenient in this case, let's roll with the `error` approach anyways.  :beers: 

(BTW I don't think I'm more experienced than you :joy_cat: )